### PR TITLE
V4.0.0 Unity

### PR DESCRIPTION
**Pull request by joeabbey**
The unity upgrade handler.

### CODE REVIEW

File: app/app.go [hunk 1]
```diff
@@ -568,6 +569,9 @@ func New(
 	app.MountTransientStores(tkeys)
 	app.MountMemoryStores(memKeys)
 
+	//regiter upgrade
```
Comment (JakeHartnell) [inline app/app.go:572]:
```suggestion
	// register upgrade
```

File: app/upgrade/upgrade_handler.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+	//get Unity Contract Address and send coin to this address
+	destAcc, _ := sdk.AccAddressFromHex(UnityContractByteAddress)
+	err := bank.SendCoins(ctx, accAddr, destAcc, sdk.NewCoins(accCoin))
+	if err != nil {
+		panic(err)
+	}
+}
+
+//CreateUpgradeHandler make upgrade handler
```
Comment (JakeHartnell) [inline app/upgrade/upgrade_handler.go:89]:
```suggestion
// CreateUpgradeHandler make upgrade handler
```

File: app/upgrade/upgrade_handler.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+		_, err := staking.CompleteUnbonding(ctx, accAddr, validatorValAddr)
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	// account balance after finishing unbonding
+	accCoin := bank.GetBalance(ctx, accAddr, bondDenom)
+
+	//get Unity Contract Address and send coin to this address
```
Comment (JakeHartnell) [inline app/upgrade/upgrade_handler.go:81]:
```suggestion
	// get Unity Contract Address and send coin to this address
```

File: app/upgrade/upgrade_test.go [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+				// move all juno from acc to community pool (uncluding bonded juno)
+				unity.ClawbackCoinFromAccount(suite.ctx, addr2, &suite.app.StakingKeeper, &bankBaseKeeper)
+
+				suite.app.EndBlock(abci.RequestEndBlock{})
+				suite.app.Commit()
+			},
+			func() {
+				//1. check if fund is moved from unbond and bond pool to community pool
+				//acc2 is supposed to lose bondTokens amount to community pool
```
Comment (JakeHartnell) [inline app/upgrade/upgrade_test.go:101]:
```suggestion
				// 1. check if fund is moved from unbond and bond pool to community pool
				// acc2 is supposed to lose bondTokens amount to community pool
```

File: app/upgrade/upgrade_test.go [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+				afterBondPool := suite.app.BankKeeper.GetBalance(suite.ctx, suite.app.StakingKeeper.GetBondedPool(suite.ctx).GetAddress(), "stake").Amount
+				afterUnbondPool := suite.app.BankKeeper.GetBalance(suite.ctx, suite.app.StakingKeeper.GetNotBondedPool(suite.ctx).GetAddress(), "stake").Amount
+
+				initial := initialBondPool.Add(initialUnbondPool)
+				later := afterBondPool.Add(afterUnbondPool)
+				require.Equal(suite.T(), initial.Sub(bondTokens), later)
+
+				//2. check if acc2 has 0 juno
```
Comment (JakeHartnell) [inline app/upgrade/upgrade_test.go:109]:
```suggestion
				// 2. check if acc2 has 0 juno
```

File: app/upgrade/upgrade_test.go [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+				afterAcc2Amount := suite.app.BankKeeper.GetBalance(suite.ctx, addr2, sdk.DefaultBondDenom).Amount
+				require.Equal(suite.T(), sdk.ZeroInt(), afterAcc2Amount)
+
+				//3. check if all unbonding delegations are removed
```
Comment (JakeHartnell) [inline app/upgrade/upgrade_test.go:113]:
```suggestion
				// 3. check if all unbonding delegations are removed
```

File: app/upgrade/constants.go [hunk 7]
```diff
@@ -0,0 +1,4 @@
+package unity
+
+//UpgradeName is upgrade name in proposal
```
Comment (JakeHartnell) [inline app/upgrade/constants.go:3]:
```suggestion
// UpgradeName is upgrade name in proposal
```

**Review by JakeHartnell — APPROVED** [review]
:shipit:

**Review by faddat — APPROVED** [review]

**Comment by giansalex** [comment]
it seems that the required approvals should be increased to 3.

**Comment by JakeHartnell** [comment]
> it seems that the required approvals should be increased to 3.

Updated. Also made it so stale PR reviews are dismissed.

**Pull request merged**